go/vt/vtgate/evalengine/expr_tuple_bvar.go

```diff
@@ -0,0 +1,110 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package evalengine
+
+import (
+	"vitess.io/vitess/go/mysql/collations"
+	"vitess.io/vitess/go/sqltypes"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+type (
+	TupleBindVariable struct {
+		Key string
+
+		Index     int
+		Type      sqltypes.Type
+		Collation collations.ID
+
+		// dynamicTypeOffset is set when the type of this bind variable cannot be calculated
+		// at translation time. Since expressions with dynamic types cannot be compiled ahead of time,
+		// compilation will be delayed until the expression is first executed with the bind variables
+		// sent by the user. See: UntypedExpr
+		dynamicTypeOffset int
+	}
+)
+
+var _ IR = (*TupleBindVariable)(nil)
+var _ Expr = (*TupleBindVariable)(nil)
+
+func (bv *TupleBindVariable) IR() IR {
+	return bv
+}
+
+func (bv *TupleBindVariable) IsExpr() {}
+
+// eval implements the expression interface
+func (bv *TupleBindVariable) eval(env *ExpressionEnv) (eval, error) {
+	bvar, err := env.lookupBindVar(bv.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	if bvar.Type != sqltypes.Tuple {
+		return nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "query argument '%s' must be a tuple (is %s)", bv.Key, bvar.Type)
+	}
+
+	tuple := make([]eval, 0, len(bvar.Values))
+	for _, value := range bvar.Values {
+		if value.Type != sqltypes.Tuple {
+			return nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "result value must be a tuple (is %s)", value.Type)
+		}
+		sValue := sqltypes.ProtoToValue(value)
+		var evalErr error
+		idx := 0
+		found := false
+		loopErr := sValue.ForEachValue(func(val sqltypes.Value) {
+			idx++
+			if idx-1 != bv.Index {
+				return
+			}
+			found = true
+			e, err := valueToEval(val, typedCoercionCollation(val.Type(), collations.CollationForType(val.Type(), bv.Collation)))
+			if err != nil {
+				evalErr = err
+				return
+			}
+			tuple = append(tuple, e)
+
+		})
+		if loopErr != nil {
+			return nil, loopErr
+		}
+		if evalErr != nil {
+			return nil, evalErr
+		}
+		if !found {
+			return nil, vterrors.VT13001("value not found in bind variable")
+		}
+	}
+	return &evalTuple{t: tuple}, nil
+}
+
+// typeof implements the expression interface
+func (bv *TupleBindVariable) typeof(env *ExpressionEnv) (ctype, error) {
+	_, err := env.lookupBindVar(bv.Key)
+	if err != nil {
+		return ctype{}, err
+	}
+
+	return ctype{Type: sqltypes.Tuple}, nil
+}
+
+func (bv *TupleBindVariable) compile(c *compiler) (ctype, error) {
+	return ctype{}, c.unsupported(bv)
```

Comment:
Can we add the logic for this here? We need this also for the type resolving as well, since we depend on compile for resolving types now as well. See for example also here:

https://github.com/vitessio/vitess/blob/c53420174d46ec5447d31deed014ee583f270e30/go/vt/vtgate/evalengine/translate.go#L697-L703

Reply:
This is a `Tuple` type bind variable and I am not sure how to compile such.
I checked `expr_bvar` and it is not recognized there
```
func (bvar *BindVariable) compile(c *compiler) (ctype, error) {
.
.
.
switch tt := typ.Type; {
.
.
default:
    return ctype{}, vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "Type is not supported: %s", tt)
}
.		
```

Reply:
We store something like
(1, 'a'), (2, 'b'), (3, 'c')... in a single bind variable.
and TupleBindVariable will provide single tuple for a index. If Index is `1`
then it return
('a'), ('b'), ('c'). The output type is still a tuple.